hudi-common/src/main/java/org/apache/hudi/common/fs/HoodieWrapperFileSystem.java

```diff
@@ -192,76 +233,110 @@ public FSDataOutputStream create(Path f, FsPermission permission, boolean overwr
+    TimedSizeAwareOutputStream tos = new TimedSizeAwareOutputStream(path, fsDataOutputStream, consistencyGuard,
         () -> openStreams.remove(path.getName()));
-    openStreams.put(path.getName(), os);
-    return os;
+    BufferedSizeAwareOutputStream bos = new BufferedSizeAwareOutputStream(tos, finalBufferSize);
+
+    openStreams.put(path.getName(), Pair.of(tos, bos));
+
+    return new FSDataOutputStream(bos);
   }
 
-  private FSDataInputStream wrapInputStream(final Path path, FSDataInputStream fsDataInputStream) throws IOException {
-    if (fsDataInputStream instanceof TimedFSDataInputStream) {
-      return fsDataInputStream;
-    }
-    return new TimedFSDataInputStream(path, fsDataInputStream);
+  /**
+   * The stream hierarchy after wrapping will be as follows.
+   *
+   * FSDataInputStream (returned)
+   *    BufferedFSInputStream (required for buffering)
+   *      TimedFSInputStream  (required for tracking metrics and timings)
+   *         fs.open()  (Original stream returned from underlying FileSystem)
+   *
+   * @param path
+   * @param fsDataInputStream
+   * @param bufferSize
+   * @return
+   * @throws IOException
+   */
+
+  private FSDataInputStream wrapInputStream(final Path path, FSDataInputStream fsDataInputStream,
+      int bufferSize) throws IOException {
+    LOG.info("Using buffersize=" + bufferSize + " to wrap input stream for path " + path);
+    TimedFSInputStream tis = new TimedFSInputStream(path, fsDataInputStream);
+
+    int minBufferSize = (HoodieFileFormat.isBaseFile(path) || FSUtils.isLogFile(path)) ? MIN_DATA_FILE_IO_BUFFER_SIZE : MIN_IO_BUFFER_SIZE;
+    int finalBufferSize = Math.max(minBufferSize, bufferSize);
+
+    LOG.info("Using buffersize=" + finalBufferSize + " to wrap output stream for path " + path);
+    return new FSDataInputStream(new BufferedFSInputStream(tis, finalBufferSize));
   }
 
   @Override
   public FSDataOutputStream create(Path f, boolean overwrite) throws IOException {
     return executeFuncWithTimeMetrics(MetricName.create.name(), f, () -> {
-      return wrapOutputStream(f, fileSystem.create(convertToDefaultPath(f), overwrite));
+      return wrapOutputStream(f, fileSystem.create(convertToDefaultPath(f), overwrite), 0);
     });
   }
 
   @Override
   public FSDataOutputStream create(Path f) throws IOException {
     return executeFuncWithTimeMetrics(MetricName.create.name(), f, () -> {
-      return wrapOutputStream(f, fileSystem.create(convertToDefaultPath(f)));
+      return wrapOutputStream(f, fileSystem.create(convertToDefaultPath(f)), 0);
     });
   }
 
   @Override
   public FSDataOutputStream create(Path f, Progressable progress) throws IOException {
     return executeFuncWithTimeMetrics(MetricName.create.name(), f, () -> {
-      return wrapOutputStream(f, fileSystem.create(convertToDefaultPath(f), progress));
+      return wrapOutputStream(f, fileSystem.create(convertToDefaultPath(f), progress), 0);
```

Comment:
create a overloaded method for `wrapOutputStream(..)` that passes `0` for the last argument?

Reply:
Done